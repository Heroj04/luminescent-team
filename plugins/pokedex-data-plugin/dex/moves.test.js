import {
  generateMovesViaLearnset,
  getMoveId,
  getMoveString,
  isMoveNameSmogonCompatible,
  getEggMoves,
  getTechMachineLearnset,
  getMoveProperties,
  getPokemonLearnset,
  parseTmLearnsetSection,
} from './moves';

describe('Dex Utils Move Getters', () => {
  describe('generateMovesViaLearnset', () => {
    it('should return the expected moves for Pikachu at level 20', () => {
      const monsNo = 25; // Pikachu
      const level = 20;
      const expectedMoves = ['Electro Ball', 'Shock Wave', 'Feint', 'Spark'];
      const actualMoves = generateMovesViaLearnset(monsNo, level);
      expect(actualMoves).toEqual(expectedMoves);
    });

    it('should return the expected moves for Pikachu at level 15', () => {
      const monsNo = 25; // Pikachu
      const level = 15;
      const expectedMoves = ['Thunder Wave', 'Double Kick', 'Electro Ball', 'Shock Wave'];
      const actualMoves = generateMovesViaLearnset(monsNo, level);
      expect(actualMoves).toEqual(expectedMoves);
    });

    it('should return the expected moves for Bulbasaur at level 10', () => {
      const bulbasaurNo = 1;
      const expectedBulbasaurMoves = ['Growl', 'Tackle', 'Vine Whip', 'Leech Seed'];
      const actualMoves = generateMovesViaLearnset(bulbasaurNo, 10);
      expect(actualMoves).toEqual(expectedBulbasaurMoves);
    });

    describe('when given invalid arguments', () => {
      it.each([
        ['invalid monsNo', 'a', 1, 'Invalid Pokémon number'],
        ['negative monsNo', -1, 1, 'Invalid Pokémon number'],
        ['monsNo not in learnset', 2899, 1, 'Invalid Pokémon number'],
        ['invalid level', 1, 'a', 'Invalid level'],
        ['negative level', 1, -1, 'Invalid level'],
      ])('should throw an error with %s', (_scenario, monsno, level, expectedError) => {
        expect(() => generateMovesViaLearnset(monsno, level)).toThrow(expectedError);
      });
    });
  });
  describe('getMoveId', () => {
    it.each([
      ['Tackle', 33],
      ['Thunderbolt', 85],
      ['Fire Blast', 126],
    ])('should return the index of the move in the moveEnum array', (input, expected) => {
      expect(getMoveId(input)).toBe(expected);
    });

    it.each([['foo'], [123], [null]])(
      'should throw an error if the move name is not found in the moveEnum array',
      (input) => {
        expect(() => getMoveId(input)).toThrowError(`Bad move name: ${input}`);
      },
    );
  });
  describe('isMoveNameSmogonCompatible', () => {
    test.each([
      [null, `Bad move string: ${null}`],
      [undefined, `Bad move string: ${undefined}`],
      [123, `Bad move string: 123`],
      ['', 'Bad move string: '],
    ])('throws %s when given %s', (moveString, expected) => {
      expect(() => isMoveNameSmogonCompatible(moveString)).toThrow(expected);
    });

    test.each([
      ['swordsdance', false],
      ['Toxic Spikes', true],
      ['Stealth Rock', true],
    ])('returns %s when given %s', (moveString, expected) => {
      expect(isMoveNameSmogonCompatible(moveString)).toBe(expected);
    });
  });

  describe('getMoveString', () => {
    it('returns the correct move string for a valid id', () => {
      expect(getMoveString(1)).toBe('Pound');
      expect(getMoveString(2)).toBe('Karate Chop');
      expect(getMoveString(3)).toBe('Double Slap');
    });

    it('throws an error for an invalid id', () => {
      const DUMMY_DATA = null;
      expect(() => getMoveString(DUMMY_DATA)).toThrow(Error(`Bad move string found: ID - ${DUMMY_DATA}`));
    });

    it('throws an error for an incompatible move string', () => {
      const incompatibleMoveId = 50000;
      const incompatibleMoveName = 'undefined';
      expect(() => getMoveString(incompatibleMoveId)).toThrow(
        Error(`Incompatible move string found: ID - ${incompatibleMoveId}, String: ${incompatibleMoveName}`),
      );
    });
  });
  describe('getEggMoves', () => {
    it('returns an empty array for an invalid dexId', () => {
      expect(getEggMoves(9999)).toEqual([]);
    });

    it('returns an array of egg moves for a valid dexId', () => {
      const eggMoves = getEggMoves(1);
      expect(eggMoves).toContainEqual({ level: 'egg', moveId: 130 });
    });

    it('returns an empty array for a dexId with no egg moves', () => {
      expect(getEggMoves(2)).toEqual([]);
    });
  });
  describe('parseTmLearnsetSection', () => {
    it('returns the correct binary string for a decimal number', () => {
      expect(parseTmLearnsetSection(2150467360)).toBe('00000100111000011011010000000001');
      expect(parseTmLearnsetSection(3149832)).toBe('00000000000001000000001000000011');
      expect(parseTmLearnsetSection(2418017312)).toBe('00000100001000000000010000001001');
    });

    it('returns a 32-character string for any input', () => {
      const binaryString = parseTmLearnsetSection(8);
      expect(binaryString).toHaveLength(32);
    });
  });
  describe('getTechMachineLearnset', () => {
    it('returns an empty array when no TMs are learned', () => {
      expect(getTechMachineLearnset(0, 0, 0, 0)).toEqual([]);
    });

    it('returns an array of TM moves when one or more TMs are learned', () => {
      const learnset = getTechMachineLearnset(2150467360, 3149832, 2418017312, 0);
      expect(learnset).toContainEqual({ level: 'tm', moveId: 92 });
      expect(learnset).toContainEqual({ level: 'tm', moveId: 331 });
      expect(learnset).toHaveLength(19);
    });

    it('ignores TMs that the Pokémon cannot learn', () => {
      const learnset = getTechMachineLearnset(2150467360, 3149832, 2418017312, 0);
      expect(learnset).not.toContainEqual({ level: 'tm', moveId: 3 });
      expect(learnset).not.toContainEqual({ level: 'tm', moveId: 4 });
    });
  });

  describe('getPokemonLearnset', () => {
    const invalidArgs = [undefined, null, -1, 'foo'];
    const validArgs = [25, 201]; // Pikachu, Unown

    it.each(invalidArgs)('returns an empty array when the argument is not a non-negative integer', (arg) => {
      expect(getPokemonLearnset(arg)).toEqual([]);
    });

    it.each(validArgs)('returns an array of moves when the argument is a valid Pokemon ID', (arg) => {
      const learnset = getPokemonLearnset(arg);
      expect(Array.isArray(learnset)).toBe(true);
    });

    it('returns an empty array when the argument is a valid Pokemon ID with no learnset', () => {
      expect(getPokemonLearnset(0)).toEqual([]);
    });
  });

  describe('getMoveProperties', () => {
    test.each([
      [1, 'Pound', 'The target is physically pounded with a long tail, a foreleg, or the like.', 0, 1, 56, 40, 100],
      [
        100,
        'Teleport',
        'The user switches places with another party Pokémon. It may also be used to warp to the last Pokémon Center visited. If a wild Pokémon uses this move, it flees.',
        13,
        0,
        32,
        0,
        101,
      ],
      [
        250,
        'Whirlpool',
        'The user traps the target in a violent swirling whirlpool for four to five turns.',
        10,
        2,
        24,
        35,
        85,
      ],
    ])(
      'returns the expected properties for moveId $moveId',
      (moveId, name, desc, type, damageType, maxPP, power, accuracy) => {
        expect(getMoveProperties(moveId)).toEqual({
          name,
          desc,
          type,
          damageType,
          maxPP,
          power,
          accuracy,
        });
      },
    );

    test('returns default values for an invalid moveId', () => {
      expect(getMoveProperties()).toEqual({
        name: '———',
        desc: '',
        type: 0,
        damageType: 0,
        maxPP: 0,
        power: 0,
        accuracy: 0,
      });
    });
  });
});
