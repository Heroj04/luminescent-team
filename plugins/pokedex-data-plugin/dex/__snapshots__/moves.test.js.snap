// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 1`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 100,
    "damageType": 1,
    "desc": "The user tucks in its head to raise its Defense stat on the first turn, then rams the target on the next turn.",
    "maxPP": 16,
    "moveId": 130,
    "name": "Skull Bash",
    "power": 130,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 2`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "The user attacks the target by scattering petals for two to three turns. The user then becomes confused.",
    "maxPP": 16,
    "moveId": 80,
    "name": "Petal Dance",
    "power": 120,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 3`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "A move that works differently for the Ghost type than for all other types.",
    "maxPP": 16,
    "moveId": 174,
    "name": "Curse",
    "power": 0,
    "type": 7,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 4`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "The user lays roots that restore its HP on every turn. Because it’s rooted, it can’t switch out.",
    "maxPP": 32,
    "moveId": 275,
    "name": "Ingrain",
    "power": 0,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 5`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "This attack makes use of nature’s power. Its effects vary depending on the user’s environment.",
    "maxPP": 32,
    "moveId": 267,
    "name": "Nature Power",
    "power": 0,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 6`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 100,
    "damageType": 0,
    "desc": "The user gazes at the target rather charmingly, making it less wary. This harshly lowers the target’s Attack stat.",
    "maxPP": 32,
    "moveId": 204,
    "name": "Charm",
    "power": 0,
    "type": 17,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 7`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 101,
    "damageType": 2,
    "desc": "The user scatters curious leaves that chase the target. This attack never misses.",
    "maxPP": 32,
    "moveId": 345,
    "name": "Magical Leaf",
    "power": 60,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 8`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "The user temporarily empties its mind to forget its concerns. This sharply raises the user’s Sp. Def stat.",
    "maxPP": 32,
    "moveId": 133,
    "name": "Amnesia",
    "power": 0,
    "type": 13,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 9`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 90,
    "damageType": 2,
    "desc": "The user whips up a storm of leaves around the target. The attack’s recoil harshly lowers the user’s Sp. Atk stat.",
    "maxPP": 8,
    "moveId": 437,
    "name": "Leaf Storm",
    "power": 130,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 10`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 85,
    "damageType": 1,
    "desc": "The user violently whirls its vines, tentacles, or the like to harshly lash the target.",
    "maxPP": 16,
    "moveId": 438,
    "name": "Power Whip",
    "power": 120,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 11`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "Unsanitary sludge is hurled at the target. This may also poison the target.",
    "maxPP": 24,
    "moveId": 124,
    "name": "Sludge",
    "power": 75,
    "type": 3,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 12`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "The user turns the ground to grass for five turns. This restores the HP of Pokémon on the ground a little every turn and powers up Grass-type moves.",
    "maxPP": 16,
    "moveId": 580,
    "name": "Grassy Terrain",
    "power": 0,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 13`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 55,
    "damageType": 0,
    "desc": "The user plays a pleasant melody that lulls the target into a deep sleep.",
    "maxPP": 24,
    "moveId": 320,
    "name": "Grass Whistle",
    "power": 0,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getEggMoves returns an array of egg moves for a valid dexId 14`] = `
{
  "level": "egg",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "In this two-turn attack, blades of wind hit opposing Pokémon on the second turn. Critical hits land more easily.",
    "maxPP": 16,
    "moveId": 13,
    "name": "Razor Wind",
    "power": 80,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 1`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 90,
    "damageType": 0,
    "desc": "A move that leaves the target badly poisoned. Its poison damage worsens every turn.",
    "maxPP": 16,
    "moveId": 92,
    "name": "Toxic",
    "power": 0,
    "type": 3,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 2`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 1,
    "desc": "The user forcefully shoots seeds at the target two to five times in a row.",
    "maxPP": 48,
    "moveId": 331,
    "name": "Bullet Seed",
    "power": 25,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 3`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "A unique attack that varies in type depending on the Pokémon using it.",
    "maxPP": 24,
    "moveId": 237,
    "name": "Hidden Power",
    "power": 60,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 4`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "The user intensifies the sun for five turns, powering up Fire-type moves. It lowers the power of Water-type moves.",
    "maxPP": 8,
    "moveId": 241,
    "name": "Sunny Day",
    "power": 0,
    "type": 9,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 5`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "A wondrous wall of light is put up to reduce damage from special attacks for five turns.",
    "maxPP": 48,
    "moveId": 113,
    "name": "Light Screen",
    "power": 0,
    "type": 13,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 6`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "This move enables the user to protect itself from all attacks. Its chance of failing rises if it is used in succession.",
    "maxPP": 16,
    "moveId": 182,
    "name": "Protect",
    "power": 0,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 7`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "A nutrient-draining attack. The user’s HP is restored by half the damage taken by the target.",
    "maxPP": 16,
    "moveId": 202,
    "name": "Giga Drain",
    "power": 75,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 8`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "The user creates a protective field that prevents status conditions for five turns.",
    "maxPP": 40,
    "moveId": 219,
    "name": "Safeguard",
    "power": 0,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 9`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "In this two-turn attack, the user gathers light, then blasts a bundled beam on the next turn.",
    "maxPP": 16,
    "moveId": 76,
    "name": "Solar Beam",
    "power": 120,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 10`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "By moving rapidly, the user makes illusory copies of itself to raise its evasiveness.",
    "maxPP": 24,
    "moveId": 104,
    "name": "Double Team",
    "power": 0,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 11`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 1,
    "desc": "The user attacks and steals the target’s held item simultaneously. The user can’t steal anything if it already holds an item.",
    "maxPP": 40,
    "moveId": 168,
    "name": "Thief",
    "power": 60,
    "type": 16,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 12`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "The user shoots boiling hot water at its target. This may also leave the target with a burn.",
    "maxPP": 24,
    "moveId": 503,
    "name": "Scald",
    "power": 80,
    "type": 10,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 13`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "The user stimulates its brain by thinking bad thoughts. This sharply raises the user’s Sp. Atk stat.",
    "maxPP": 32,
    "moveId": 417,
    "name": "Nasty Plot",
    "power": 0,
    "type": 16,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 14`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 1,
    "desc": "The user attacks everything around it by causing a tremendous explosion. The user faints upon using this move.",
    "maxPP": 8,
    "moveId": 153,
    "name": "Explosion",
    "power": 250,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 15`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 0,
    "desc": "The user flashes a light that lowers the target’s accuracy. It can also be used to illuminate caves.",
    "maxPP": 32,
    "moveId": 148,
    "name": "Flash",
    "power": 0,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 16`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 101,
    "damageType": 0,
    "desc": "A frenetic dance to uplift the fighting spirit. This sharply raises the user’s Attack stat.",
    "maxPP": 32,
    "moveId": 14,
    "name": "Swords Dance",
    "power": 0,
    "type": 0,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 17`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 2,
    "desc": "The user snares the target with grass and trips it. The heavier the target, the greater the move’s power.",
    "maxPP": 32,
    "moveId": 447,
    "name": "Grass Knot",
    "power": 1,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 18`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 1,
    "desc": "The target is cut with a scythe or claw. It’s also one of the Pokétch’s hidden moves.",
    "maxPP": 40,
    "moveId": 15,
    "name": "Cut",
    "power": 60,
    "type": 11,
  },
}
`;

exports[`Dex Utils Move Getters getTechMachineLearnset returns an array of TM moves when one or more TMs are learned 19`] = `
{
  "level": "tm",
  "move": {
    "accuracy": 100,
    "damageType": 1,
    "desc": "The target is slugged with a punch thrown at maximum power. It’s also one of the Pokétch’s hidden moves.",
    "maxPP": 24,
    "moveId": 70,
    "name": "Strength",
    "power": 100,
    "type": 0,
  },
}
`;
